new simulated board: nrf52_bsim

aescolar opened the pull request:
new simulated board: nrf52_bsim 
Added a new simulated board, which models the NRF52832 SOC.

Its main use is for workstation testing and simulation of
the BLE stack and any application which relies mostly on it.

It uses BabbleSim (http://Babblesim.github.io) for the radio
simulation.
And the NRF52 HW models hosted in that same GitHub organization:
https://github.com/BabbleSim/ext_NRF52_hw_models

For speed it uses the POSIX arch to (not) emulate the CPU.
It uses Vanilla Zephyr, with a couple of configuration differences:
* It uses (like other POSIX arch boards) the system libC
* It does not use the nrfx hosted by Zehpyr in ext/, but the one
  provided by the HW models.
Otherwise it relies in the same drivers as the real NRF52 boards.
--------------------------

#### Quick hands on:
Assuming you want to have BabbleSim in ${HOME}/playground/bsim
```
mkdir -p ${HOME}/playground/bsim && cd ${HOME}/playground/bsim
# You need android's repo ( https://source.android.com/setup/develop/repo ) 
# Nordic guys: You need ssh access to GitHub.. (if you can't, do "repo init -u https://github.com/BabbleSim/manifest.git -m zephyr_ci.xml -b master")
repo init -u git@github.com:BabbleSim/manifest.git -m everything.xml -b master
repo sync
make everything -j 8
# patiently wait 1min
```
Now, you just go to Zephyr (pick this branch), and before compiling or running sanitycheck you set
```
export BSIM_OUT_PATH=${HOME}/playground/bsim/
export BSIM_COMPONENTS_PATH=${HOME}/playground/bsim/components/
```
You are good to go. Compile whatever you want targeting `nrf52_bsim`
The resulting exe can just be run stand alone without options (and you get a warning about not connecting to the BabbleSim phy)

If you want to see BLE activity (a bit more fun) you can either build each by hand and copy it out to the `bsim/bin` directory given them a reasonable name, or use a silly Makefile I left in `bsim/components/ext_zephyr_app_builder/`.
Assuming the second:
```
export BOARD=nrf52_bsim
export ZEPHYR_BASE=<wherever you have Zephyr>
cd ~/playground/bsim/components/ext_zephyr_app_builder/
make APP=samples/bluetooth/central_hr install
make APP=samples/bluetooth/peripheral install
```
And you run them:
```
cd ~/playground/bsim/bin/
./bs_nrf52_bsim_samples_bluetooth_peripheral -s=Hola -d=0 &
./bs_nrf52_bsim_samples_bluetooth_central_hr -s=Hola -d=1 &
./bs_2G4_phy_v1 -s=Hola -D=2 -sim_length=10e6 &
```
You should get something like so:
```
d_00: @00:00:00.000000  ***** Booting Zephyr OS v1.9.0-7263-g0affa09 *****
d_00: @00:00:00.000000  [bt] [INF] hci_vs_init: HW Platform: Nordic Semiconductor (0x0002)
d_00: @00:00:00.000000  [bt] [INF] hci_vs_init: HW Variant: nRF52x (0x0002)
d_00: @00:00:00.000000  [bt] [INF] hci_vs_init: Firmware: Standard Bluetooth controller (0x00) Version 1.13 Build 99
d_00: @00:00:00.000000  [bt] [WRN] bt_setup_id_addr: No static addresses stored in controller
d_01: @00:00:00.000000  ***** Booting Zephyr OS v1.9.0-7263-g0affa09 *****
d_00: @00:00:00.002528  [bt] [INF] bt_dev_show_info: Identity: e0:3b:58:29:c3:12 (random)
d_00: @00:00:00.002528  [bt] [INF] bt_dev_show_info: HCI: version 5.0 (0x09) revision 0x0000, manufacturer 0x05f1
d_00: @00:00:00.002528  [bt] [INF] bt_dev_show_info: LMP: version 5.0 (0x09) subver 0xffff
d_01: @00:00:00.000000  [bt] [INF] hci_vs_init: HW Platform: Nordic Semiconductor (0x0002)
d_01: @00:00:00.000000  [bt] [INF] hci_vs_init: HW Variant: nRF52x (0x0002)
d_01: @00:00:00.000000  [bt] [INF] hci_vs_init: Firmware: Standard Bluetooth controller (0x00) Version 1.13 Build 99
d_00: @00:00:00.004448  [bt] [WRN] irk_init: Using temporary IRK
d_00: @00:00:00.004448  Bluetooth initialized
d_01: @00:00:00.000000  [bt] [WRN] bt_setup_id_addr: No static addresses stored in controller
d_00: @00:00:00.005048  Advertising successfully started
d_01: @00:00:00.002528  [bt] [INF] bt_dev_show_info: Identity: cf:71:44:b1:06:6e (random)
d_01: @00:00:00.002528  [bt] [INF] bt_dev_show_info: HCI: version 5.0 (0x09) revision 0x0000, manufacturer 0x05f1
d_01: @00:00:00.002528  [bt] [INF] bt_dev_show_info: LMP: version 5.0 (0x09) subver 0xffff
d_01: @00:00:00.002528  Bluetooth initialized
d_01: @00:00:00.003488  Scanning successfully started
d_01: @00:00:00.007009  [DEVICE]: 7a:2c:95:7a:2c:95 (random), AD evt type 0, AD data len 29, RSSI -35
d_01: @00:00:00.007009  [AD]: 1 data_len 1
d_01: @00:00:00.007009  [AD]: 3 data_len 6
d_01: @00:00:00.115382  Connected: 7a:2c:95:7a:2c:95 (random)
d_00: @00:00:00.115884  Connected
d_01: @00:00:00.118849  [ATTRIBUTE] handle 10
d_01: @00:00:00.168728  [ATTRIBUTE] handle 11
d_01: @00:00:00.218848  [ATTRIBUTE] handle 13
d_01: @00:00:00.218848  [SUBSCRIBED]
d_01: @00:00:01.017424  [NOTIFICATION] data 0x080a34ab length 2
d_01: @00:00:02.017424  [NOTIFICATION] data 0x080a34ab length 2
[..]
d_01: @00:00:08.067424  [NOTIFICATION] data 0x080a34ab length 2
d_01: @00:00:09.117424  [NOTIFICATION] data 0x080a34ab length 2
```
Now you can go wild and run either (or both) in gdb, valgrind or whatever you fancy.

aescolar committed:
tests: lib: mem_alloc: Blacklist all POSIX arch boards

These tests rely on compiling with one of the provided
C libraries which are not compiled for any
of the POSIX ARCH boards => balcklist based on the arch
and not on the board

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
drivers: console: native: header fix

posix_flush_stdout() must be provided by any board
using CONFIG_NATIVE_POSIX_CONSOLE, not just by those using
CONFIG_NATIVE_POSIX_STDOUT_CONSOLE

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
new simulated board: nrf52_bsim

Added a new simulated board, which models the NRF52832 SOC.

Its main use is for workstation testing and simulation of
the BLE stack and any application which relies mostly on it.

It uses BabbleSim (http://Babblesim.github.io) for the radio
simulation.
And the NRF52 HW models hosted in that same GitHub organization:
https://github.com/BabbleSim/ext_NRF52_hw_models

For speed it uses the POSIX arch to (not) emulate the CPU.
It uses Vanilla Zephyr, with a couple of configuration differences:
* It uses (like other POSIX arch boards) the system libC
* It does not use the nrfx hosted by Zehpyr in ext/, but the one
  provided by the HW models.
Otherwise it relies in the same drivers as the real NRF52 boards.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar commented:
@dbkinder : The documentation cannot be built yet, I need to fix the kconfig of this board (it has quite a few hacks that were fine enough when the board was off-tree, but are not ok when merging it in)

aescolar commented:
@nashif : If we run coverage in native_posix and this board we have too many tests with coverage and they spill to the 2nd CI instance (at least in this run with 7 instances). As we only collect coverage data from the 1st instance, that would lead to the coverage reports jumping up and down.

aescolar commented:
The coverage results are here: https://codecov.io/gh/zephyrproject-rtos/zephyr/pull/9988

dbkinder commented on boards/posix/nrf52_bsim/doc/board.rst:
```diff
@@ -0,0 +1,186 @@
+* Clock control
+* PPI (Programmable Peripheral Interconnect)
+
+This board utilizes the POSIX arch to (not) simulate the CPU.
```
Better would be:

    The nrf52_bsim board definition uses the POSIX architecture to
    run applications natively on the development system.  As with
    the ``native_posix`` board, this has the benefit of providing 
    native code execution performance and easy debugging using
    native tools, but has the same drawbacks.  Please refer to
    :ref:`Native Posix's important limitations <Important_limitations>`
    for more details.

dbkinder commented on boards/posix/nrf52_bsim/doc/board.rst:
```diff
@@ -0,0 +1,186 @@
+   :goals: build
+   :compact:
+
+Now you can execute your application as:
```
change to:

    Then you can execute your application using:

dbkinder commented on boards/posix/nrf52_bsim/doc/board.rst:
```diff
@@ -0,0 +1,186 @@
+`Shell support`_ section.
+section for more information.
+
+.. _`Shell support`:
```
and delete these two lines

aescolar commented on boards/posix/native_posix/doc/board.rst:
```diff
@@ -35,6 +35,8 @@ It has only been tested on Linux, but should also be compatible with macOS.
    This port will **not** work in Windows Subsystem for Linux (WSL) because WSL
    does not support native 32-bit binaries.
 
+.. _native_important_limitations:
```
@dbkinder : Are these 2 extra references needed after your changes? Or should I just delete them?

aescolar commented on boards/posix/nrf52_bsim/doc/board.rst:
```diff
@@ -0,0 +1,181 @@
+the ``native_posix`` board, this has the benefit of providing
+native code execution performance and easy debugging using
+native tools, but has the same drawbacks.  Please refer to
+:ref:`Native Posix's important limitations <Important_limitations>`
```
Thanks David, with these 2 fixes is building nicely in my machine.

aescolar commented on boards/posix/nrf52_bsim/Kconfig.defconfig:
```diff
@@ -0,0 +1,62 @@
+#but we would not get them otherwise as we do not really inherit this SOC
+#configuration (and we cannot pull the ARM related configs)
+
+#config SOC_FAMILY_NRF
```
Note that these commented configs are here only so it is evident what was needed to make it work without all the #if changes. Having these re-definitions was ok while the board was off-tree. But when bringing the board in, the changes in the "kconfig fix" commit are necessary.
This commit would be squashed into the board one, and this comments would be removed before merging.

aescolar commented:
@dbkinder : Is the documentation acceptable after the corrections, or would you like more changes?

jhedberg commented:
I've done some basic Bluetooth testing with this, and it seems to be working fine. I haven't done any thorough code-review however.

dbkinder reviewed (approved):
LGTM, thanks.

cvinayak reviewed (approved):
QUICK-ACK. LGTM.

carlescufi commented on boards/posix/nrf52_bsim/bstests.h:
```diff
@@ -0,0 +1,121 @@
+ * Will be called with the command line arguments for the testcase.
+ * This is BEFORE any SW has run, and before the HW has been initialized
+ */
+typedef void (*bst_test_args_t)(int, char**);
```
This whole file does not really follow the Zephyr coding conventions. You add a lot of typedefs and those lines seem a bit too long to me. Is there a reason for this?

aescolar replied:
All lines are under 80 chars :)
typedefs for functions types seem, luckily, allowed

carlescufi replied:
Fine, then it was just my eyes misleading me :)

aescolar commented:
Note to reviewers: This last push is a rebase on top of today's master + 2 extra commits (which before merging would be squashed on the previous, but by now I leave separated to ease reviewing)

aescolar committed:
soc: Kconfig: Add options for compatible NRF SOCs

Simulated NRF boards based on the POSIX ARCH cannot directly
use the current SOC_*NRF options as these pull options from
the ARM CPU and other peripherals.
This commit adds a new set of hidden SOC_COMPATIBLE_* options to
be selected both by the real SOCs and the simulated ones.
In this manner we can have the common code depend on the
SOC_COMPATIBLE* options instead of the current ones where
neccessary.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

(f) arm kconfig

aescolar committed:
clock_control: nRF5x: use new compatible SOC symbol

Depend on the new SOC_COMPATIBLE_NRF symbol instead of the
SOC_FAMILY_NRF symbol.

This allows the driver to be used from the bsim simulated board

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
drivers: timer nRF: simplify kconfig depend

Simplify the Kconfig dependency for the nrf timer driver.
CLOCK_CONTROL_NRF5 depends on the SOC_FAMILY_NRF already.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
drivers: entropy nRF5x : use new compatible SOC symbol

Depend on the new SOC_COMPATIBLE_NRF symbol instead of the
SOC_FAMILY_NRF symbol.

This allows the driver to be used from the bsim simulated board

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
Bluetooth: Use new nrf COMPATIBLE options

To allow the BLE stack to be used both in the real nRF platforms
and simulated ones, change the used macros in the code to the
COMPATIBLE ones.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
drivers: timer nRF: use new BabbleSim config option

The simulator symbol was renamed.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar commented:
@ulfalizer : could you take a look at the kconfig changes and tell if they seem sensible?

ulfalizer commented on boards/posix/nrf52_bsim/Kconfig:
```diff
@@ -0,0 +1,32 @@
+
+config SOC_SERIES_BSIM_NRFXX
+	bool
+	default n
```
`bool` symbols implicitly default to `n`, so `default n` can be skipped.

That generates clearer docs too:

    Implicitly defaults to n

instead of

    - n if <dependencies>

ulfalizer commented on boards/posix/nrf52_bsim/Kconfig:
```diff
@@ -0,0 +1,32 @@
+
+config SOC_SERIES_BSIM_NRF52X
+	bool
+	default n
```
Redundant `default n`.

ulfalizer commented on boards/posix/nrf52_bsim/Kconfig.defconfig:
```diff
@@ -0,0 +1,51 @@
+if BOARD_NRF52_BSIM
+
+config BUILD_OUTPUT_BIN
+	default n
```
(This one might be okay, as it could override a `default y` from the base definition. Same below.)

aescolar replied:
Yep, it is overriding the misc/Kconfig:225 default y

ulfalizer commented on boards/posix/nrf52_bsim/Kconfig:
```diff
@@ -0,0 +1,32 @@
+
+#This would eventually be shared by a possible family of simulated NRF boards
+#which use BabbleSim. When that happens, we can move this to a common
+#Kconfig file
```
Nit: Putting a space after `#` would make it consistent with the other comments.

ulfalizer commented on soc/arm/nordic_nrf/Kconfig:
```diff
@@ -18,3 +18,16 @@ source "soc/arm/nordic_nrf/Kconfig.peripherals"
+
+config SOC_COMPATIBLE_NRF52832
+	default y if SOC_NRF52832
+	bool
```
Super nit: Missing newline

ulfalizer commented on boards/posix/nrf52_bsim/Kconfig.defconfig:
```diff
@@ -0,0 +1,51 @@
+	def_bool y
+
+config SOC_COMPATIBLE_NRF52832
+	def_bool y
```
Not super familiar with the details here, but instead of having `Kconfig.defconfig` files override `SOC_COMPATIBLE_*` symbols, maybe they could be `select`ed by other `SOC_*` symbols.

`SOC_FOO_MARK_II` could have `select SOC_COMPATIBLE_FOO` for example, if it's backwards compatible.

aescolar replied:
I did try first to select them in the kconfig.board, but it did not get thru. So I went for the easy path.. 
I guess the problem is that they are defined in the real SOC kconfig files and therefore only included when the real SOC and ARM arch is selected?

ulfalizer replied:
It's fine if they become undefined for other ARCHes by the way. Undefined symbols evaluate to `n`, and the CI check for undefined symbols (in `scripts/ci/check-compliance.py`) always parses the entire Kconfig tree (it includes all boards and ARCHes by setting `BOARD_DIR` and `ARCH` to glob patterns).

ulfalizer replied:
Just noticed that the simulated boards use the POSIX arch, so that might not work. Maybe the symbols could be put into `soc/Kconfig` instead. Since non-ARM arches can be compatible with those boards, it'd still make sense from an organizational perspective.

aescolar replied:
> could be put into soc/Kconfig

It would feel funny to me to have that kind of symbol in such a high common place.. Even where they are today, seems a bit to out of place..

ulfalizer replied:
IMO it makes sense considering that the symbols are arch-independent. Putting them in `soc/Kconfig` makes it explicit.

Having a POSIX board be compatible with an ARM board is pretty unexpected in itself, so it's probably a good thing if it stands out a bit. It makes it easier to notice that something special is going on. :)

(I suspect that a lot of the `Kconfig.defconfig` stuff in Zephyr could be simplified a bit with promptless helper symbols and `select` too, but that's a different issue.)

Maybe it'd be nice to add a comment:

```diff
diff --git a/soc/Kconfig b/soc/Kconfig
index cc6a7ee14..17a3219cb 100644
--- a/soc/Kconfig
+++ b/soc/Kconfig
@@ -11,3 +11,22 @@ osource "$(SOC_DIR)/$(ARCH)/Kconfig"
 osource "$(SOC_DIR)/$(ARCH)/*/Kconfig"
 
 endmenu
+
+# The helper symbols below are put here due to an unusual setup: The simulated
+# nrf52_bsim board uses the POSIX arch, but is compatible with Nordic ARM
+# boards
+
+config SOC_COMPATIBLE_NRF
+       bool
+       help
+         Selected by NRF-compatible boards
+
+config SOC_COMPATIBLE_NRF52X
+       bool
+       help
+         Selected by NRF52X-compatible boards
+
+config SOC_COMPATIBLE_NRF52832
+       bool
+       help
+         Selected by NRF52832-compatible boards
```

aescolar replied:
Anybody disagrees with what Ulf proposes? otherwise I will just push his proposal.

ulfalizer commented on boards/posix/nrf52_bsim/Kconfig.defconfig:
```diff
@@ -0,0 +1,51 @@
+
+if LOG
+
+#This board can reuse the native_posix logging backend
```
Missing space before `#`.

ulfalizer commented on boards/posix/nrf52_bsim/Kconfig.defconfig:
```diff
@@ -0,0 +1,51 @@
+
+endif # LOG
+
+endif #BOARD_NRF52_BSIM
```
Missing space before `#`.

ulfalizer commented on boards/posix/nrf52_bsim/Kconfig.defconfig:
```diff
@@ -0,0 +1,51 @@
+config BT_CTLR
+	def_bool y
+
+endif #BT
```
Missing space before `#`.

ulfalizer reviewed (changes requested):
The `SOC_COMPATIBLE_*` helper symbol approach makes sense to me. I wonder if they could be `select`ed instead of having to set them via `Kconfig.defconfig` files though. Might be a bit simpler.

ulfalizer commented:
Hmm... If those helper symbols have non-trivial dependencies, `select` might be risky though. Judgment call there.

`select` is great for simple promptless helper symbols without dependencies at least. Usually, it's a good idea to avoid it otherwise.

carlescufi reviewed (approved):
Looks good to me with regards to the nRF integration bits.

aescolar commented:
recheck

ulfalizer reviewed (approved):
Kconfig parts look fine.

aescolar commented:
From my point of view this is ready to merge.
I just made this new board not default to avoid adding too much extra load to the CI runs.

nashif commented on soc/Kconfig:
```diff
@@ -11,3 +11,19 @@ osource "$(SOC_DIR)/$(ARCH)/Kconfig"
 
 endmenu
+
+# The helper symbols below are put here due to an unusual setup: The simulated
```
Need a better spot for this, we should not have soc specific kconfigs in a general file such as  soc/Kconfig. 
Also, I would do "select SOC_COMPATIBLE_NRF" in the SOC_FAMIL_NRF config instead of this.

aescolar replied:
> Also, I would do "select SOC_COMPATIBLE_NRF" in the SOC_FAMIL_NRF config instead of this.

Ok.

> Need a better spot for this, we should not have soc specific kconfigs in a general file such as soc/Kconfig.

Do you have a better place in mind? (please consider this comment from Ulf: https://github.com/zephyrproject-rtos/zephyr/pull/9988#discussion_r222670030 )

nashif replied:
ok, can't think of a better solution now, will revisit.

nashif commented on subsys/bluetooth/controller/CMakeLists.txt:
```diff
@@ -44,14 +44,15 @@ zephyr_library_sources_ifdef(
   hal/nrf5/ticker.c
   )
 
+
```
extra new line

nashif commented on .shippable.yml:
```diff
@@ -24,7 +24,7 @@ build:
         - ${SHIPPABLE_BUILD_DIR}/ccache
```
please split this commit into 3:
- shippable
- sanitycheck
- eveything else.

nashif commented on .shippable.yml:
```diff
@@ -34,6 +34,16 @@ build:
           git rebase origin/${PULL_REQUEST_BASE_BRANCH};
         fi
+      - >
+         export NRF_HW_MODELS_VERSION=`cat boards/posix/nrf52_bsim/hw_models_version`;
```
better define this under global section

aescolar replied:
I don't think we can have this guy defined before the rebase just above.
But I moved it into its own line to make it more visible.

nashif commented on .shippable.yml:
```diff
@@ -34,6 +34,16 @@ build:
+      - >
+         export NRF_HW_MODELS_VERSION=`cat boards/posix/nrf52_bsim/hw_models_version`;
+         pushd . ;
+         export BSIM_OUT_PATH=/opt/bsim/ ;
```
ditto

nashif commented on .shippable.yml:
```diff
@@ -34,6 +34,16 @@ build:
+         export NRF_HW_MODELS_VERSION=`cat boards/posix/nrf52_bsim/hw_models_version`;
+         pushd . ;
+         export BSIM_OUT_PATH=/opt/bsim/ ;
+         export BSIM_COMPONENTS_PATH=${BSIM_OUT_PATH}/components/;
```
ditto

aescolar committed:
nrf52_bsim: Explicitly list the applicable HW models version

To ensure we can always go back in time, let's explicitly list the
NRF52 HW models version which should be used with each version of
the nrf52_bsim board

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
sanitycheck: nf52_bsim testcases to run with other native

To be able to collect all coverage reports from the "native"
applications in the first CI instance, let's also place the
nrf52_bsim testcases first. The order now is:
 unit_testing
 native_posix
 *_bsim
 anything else

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar committed:
shippable: Add support for nrf52_bsim board

Use the docker image which has BabbleSim precompiled

Fetch the NRF52 hardware models required by the nrf52_bsim board
and compile them.

Collect the coverage report from the nrf52_bsim board

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

aescolar commented:
recheck
